protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+        distributionStarter = _starter;
+        emit DistributionStarterSet(_starter);
+    }
+
+    /**
+     * @notice Set migrator
+     * @param _migrator Migrator
+     */
+    function setMigrator(address _migrator) external override onlyElevatedAccess {
+        if (_migrator == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        migrator = _migrator;
+        emit MigratorSet(_migrator);
+    }
+
+    /**
+     * @notice Delegate votes from `msg.sender` to `delegatee`
+     * @param delegatee The address to delegate votes to
+     */
+    function delegate(address delegatee) external override {
+        return _delegate(msg.sender, delegatee);
+    }
+
+    /**
+     * @notice Set reward distribution cooldown
+     * @param _cooldown Cooldown in seconds
+     */
+    function setRewardDistributionCoolDown(uint160 _cooldown) external override onlyElevatedAccess {
+        /// @dev zero cooldown is allowed
+        rewardDistributionCoolDown = _cooldown;
+        emit RewardDistributionCoolDownSet(_cooldown);
+    }
+
+    /**
+      * @notice For migrations to a new staking contract
+      *         1. Withdraw `staker`s tokens to the new staking contract (the migrator)
+      *         2. Any existing rewards are claimed and sent directly to the `staker`
+      * @dev Called only from the new staking contract (the migrator).
+      *      `setMigrator(new_staking_contract)` needs to be called first
+      * @param staker The staker who is being migrated to a new staking contract.
+      * @return Staker balance
+      */
+    function migrateWithdraw(address staker) external override onlyMigrator returns (uint256) {
+        if (staker == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        uint256 stakerBalance = _balances[staker];
+        _withdrawFor(staker, msg.sender, stakerBalance, true, staker);
+        return stakerBalance;
+    }
+    
+    /**
+     * @notice Distributed TGLD rewards minted to this contract to stakers
+     * @dev This starts another epoch of rewards distribution. Calculates new `rewardRate` from any left over rewards up until now
+     */
+    function distributeRewards() external override whenNotPaused updateReward(address(0)) {
+        if (msg.sender != distributionStarter) { revert CommonEventsAndErrors.InvalidAccess(); }
+        if (totalSupply == 0) { revert NoStaker(); }
+        // Mint and distribute TGLD if no cooldown set
+        if (lastRewardNotificationTimestamp + rewardDistributionCoolDown > block.timestamp) 
+                { revert CannotDistribute(); }
+        _distributeGold();
+        uint256 rewardAmount = nextRewardAmount;
+        // revert if next reward is 0 or less than reward duration (final dust amounts)
+        if (rewardAmount < rewardDuration ) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        nextRewardAmount = 0;
+        _notifyReward(rewardAmount);
+        lastRewardNotificationTimestamp = uint32(block.timestamp);
```

Comment:
nit 
I would move the update of `lastRewardNotificationTimestamp` inside the `_notifyReward()`.